software/grapherrific.livecodescript

```diff
@@ -0,0 +1,1477 @@
+         -- first empty out resource information
+         -- call gravitySort outputs-to-inputs to figure total demand and increase supply to meet/exceed
+         -- call gravitySort inputs-to-outputs to figure supply/demand mismatch
+         -- keep doing that until nothing changes
+         repeat for each key tFoo in pResources[tDoc]
+            -- input & output used by old resource calculation
+            --            if tFoo is "input" then next repeat
+            --            if tFoo is "output" then next repeat
+            put empty into pResources[tDoc][tFoo]["gravity"]
+            if pShort[tDoc][tFoo]["type"] is "state" then
+               put empty into pResources[tDoc][tFoo]["excess"]
+               put empty into pResources[tDoc][tFoo]["totSupply"]
+               put empty into pResources[tDoc][tFoo]["totDemand"]
+            end if
+         end repeat
+         put true into pChanged
+         put 0 into tRepeatLimit
+         repeat while pChanged is true
+            put tRepeatLimit + 1 into tRepeatLimit
+            if tRepeatLimit > (the number of lines of the keys of pShort[tDoc]) then 
+               --               breakpoint
+               answer "error: repeat limit reached in supplyDemand function" -- ! had a bug and not sure if it's totally fixed 28NOV2015
+               exit repeat
+            end if
+            put false into pChanged -- do this repeat at least once
+            put the keys of pShort[tDoc] into pNoVisit[tDoc]
+            replace cr with "," in pNoVisit[tDoc]
+            get gravitySort(pShort,pSketch,pResources,pChanged,pNoVisit,tDoc,tDemandSide,"bulk trim prev next") -- total demand 
+            put the keys of pShort[tDoc] into pNoVisit[tDoc]
+            replace cr with "," in pNoVisit[tDoc]
+            get gravitySort(pShort,pSketch,pResources,pChanged,pNoVisit,tDoc,tSupplySide,"read skim next prev") -- supply/demand mismatch
+         end repeat
+      end if
+   end repeat -- tDoc in pShort
+end supplyDemand
+
+function gravitySort @pShort,@pSketch,@pResources,@pChanged,@pNoVisit,pDoc,pSeedList,pDirection
+   put "function gravitySort" into xX["write"];    devLog xX
+   -- sMainShort,sScopeSketch,sResources,sSomethingChanged,sNotVisited
+   -- pDoc = document number
+   -- pSeedList = ordered comma delimited list of records in pDoc
+   -- pDirection = "read skim next prev" or "bulk trim prev next"; easy to refer to as word 1/2/3/4
+   -- doesn't return anything
+   -- just walks network, edits sResources, and terminates
+   -- ! gravity logic doesn't cross scope boundaries; this balances flows but not up or down scope   
+   -- scopeRef is assigned when a node is placed into the parameter, before being sent to gravitySort
+   replace cr with "" in pSeedList -- should just be numbers and commas
+   repeat for each item tStep in pSeedList
+      --      -- ? should this be item offset of tStep
+      --      delete item 1 of pSeedList
+      set the wholeMatches to true
```

Comment:
@matthewmaier again looks like wholeMatches could just be set to true at the top of the function